File: python/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # bfs 迭代中序遍历
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        stack = list()
        in_order = float('-inf')
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if root.val <= in_order:
                return False
            in_order = root.val
            root = root.right
        return True

    # dfs 递归
    def isValidBST1(self, root: Optional[TreeNode]) -> bool:

        def helper(root, lower=float('-inf'), upper=float('inf')):
            if not root:
                return True
            if root.val >= upper or root.val <= lower:
                return False
            if not helper(root.left, lower, root.val):
                return False
            if not helper(root.right, root.val, upper):
                return False
            return True

        return helper(root)


# 迭代实现树的三种遍历 https://blog.csdn.net/Netty711/article/details/124535792

# @lc code=end

File: cpp/98.验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */
#include <iostream>
#include <stdlib.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// int INT_MAX = (2 << 31) - 1;
// int INT_MIN = 2 << 31;
long long MY_MIN_VALUE = LLONG_MIN;

class Solution {
public:
    bool isValidBST(TreeNode* root)
    {
        // return dfs(root, LLONG_MAX, LLONG_MIN);
        return dfs_math(root, MY_MIN_VALUE);
    }

    // 2024春节后再战！
    bool bfs(TreeNode* root) {

    }

    // dfs 利用二叉搜索树的特点
    bool dfs_math(TreeNode* root, long long& pre)
    {
        if (!root) {
            return true;
        }

        bool ans = true;

        if (root->left) {
            ans &= dfs_math(root->left, pre);
        }

        if (root->val > pre) {
            pre = root->val;
        } else {
            ans = false;
        }

        if (root->right) {
            ans &= dfs_math(root->right, pre);
        }

        return ans;
    }

    bool dfs(TreeNode* root, long long high, long long low)
    {
        if (!root) {
            return true;
        }

        if (root->val <= low || root->val >= high) {
            return false;
        }

        return dfs(root->left, root->val, low) && dfs(root->right, high, root->val);
    }

    long long MY_MIN_VALUE = LLONG_MIN;
};
// @lc code=end

int main()
{
    TreeNode root;
    root.val = 0;
    bool res = (Solution()).dfs_math(&root, MY_MIN_VALUE);
    cout << res << endl;
}
